rfcs/20240806-c-api/README.md

```diff
@@ -0,0 +1,128 @@
+# C API Design Document (RFC)
+
+
+## Introduction
+
+The oneCCL communication library’s current APIs is defined in the [oneAPI
+specification][ccl-spec]. However, other APIs used by similar collective
+communication libraries differ from those used by oneCCL. For example, see
+[NCCL][nccl-spec] from Nvidia, [RCCL][rccl-spec] from AMD, and hccl from
+Habana. This RFC asks for feedback about aligning the oneCCL APIs to be closer
+to other vendor libraries, since this facilitates integration with frameworks
+and upstreaming to the open source. 
+
+One difference between oneCCL and other vendors communication libraries is that
+all other communication libraries have a C API, while oneCCL has a C++ API.
+This is because oneCCL was designed to integrate with SYCL, which is based on
+C++. One of the goals of oneCCL is to support different hardware and vendors,
+such as Intel Data Center GPU Max Series, Intel Core and Intel Xeon family,
+Intel Gaudi, Nvidia or AMD GPUs, among others. 
+
+[ccl-spec]: https://uxlfoundation.github.io/oneAPI-spec/spec/elements/oneCCL/source/index.html
+[hccl-spec]: https://docs.habana.ai/en/latest/API_Reference_Guides/HCCL_APIs/C_API.html
+[nccl-spec]: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api.html
+[rccl-spec]: https://rocm.docs.amd.com/projects/rccl/en/latest/api-reference/api-library.html#api-library
+
+## Proposal
+
+The proposal is to define a C-like API that aligns with current APIs in other
+communication libraries, while introducing a few changes, as described next:
+
+1. Most APIs are C-based like other communication libraries. C++ data
+   structures are hidden behind handles returned to the user, such as
+   `ccl::stream` and `ccl::comm`. 
+
+2. The API is extended to support different types of streams or queues:
+
+   - `onecclResult_t  onecclCreateStreamXPU(onecclStream_t* oneccl_stream, void *args)`
+      the args is a pointer to the stream or queue that is vendor specific.
+   - `onecclResult_t  onecclStreamCreateCPU(onecclStream_t* oneccl_stream, void* args)`
+      this API is explicit for CPU. 
+   - `onecclResult_t onecclReleaseStream(oneccl_stream)`
+
+   `onecclResult_t onecclStreamDestroy(onecclStream_t oneccl_stream)`
+
+   Once the sycl::queue is registered, it is hidden behind the `cclStream_t`
```

Comment:
`cclStream_t` -> `onecclStream_t`

Reply:
Yes, this is a typo. @zhenggb72, can you update this and we can approve and merge?